parsl/monitoring/radios.py

```diff
@@ -173,3 +174,17 @@ def send(self, message: object) -> None:
             logging.error("Could not send message within timeout limit")
             return
         return
+
+
+class MultiprocessingQueueRadio(MonitoringRadio):
+    """A monitoring radio intended which connects over a multiprocessing Queue.
+    This radio is intended to be used on the submit side, where components
+    in the submit process, or processes launched by multiprocessing, will have
+    access to a Queue shared with the monitoring database code (bypassing the
+    monitoring router).
+    """
+    def __init__(self, queue: Queue) -> None:
+        self.queue = queue
+
+    def send(self, message: object) -> None:
+        self.queue.put((message, 0))
```

Comment:
`message` is a generic, nominally pickleable object.  Okay ... and we're wrapping into a tuple for the other end.  What is the magic number `0`?

Also, where is the end of this queue?  As in, where is it drained?

Reply:
It's the IP address this UDP packet was received from! (in the original UDP implementation of monitoring radios)

Really that should get removed from the MonitoringRadio API but not in this PR...

The other end of the queue is in the bowels of the monitoring database manager... it's super confusing to follow that code because there are many different queues all draining into each other...  it's this block_msgs queue in MonitoringHub:

``` self.radio = MultiprocessingQueueRadio(self.block_msgs)```

which is consumed in the dbm_proc process started inside the MonitoringHub, which is processed in this thread in `parsl/monitoring/db_manager.py`

```
        self._block_queue_pull_thread = threading.Thread(target=self._migrate_logs_to_internal,
                                                         args=(
                                                             block_queue, 'block', self._kill_event,),
                                                         name="Monitoring-migrate-block",
                                                         daemon=True,
                                                         )
```

The queue already exists and is passed to the router code to be populated from a ZMQ receiver and a UDP receiver too - so this radio is providing another way to get messages into that queue that doesn't use ZMQ or UDP.

In general that code has parallel queues with different names (resources, blocks, priority) because they all used to have different processing behaviour in various places. Over time, I've worked towards making them all behave the same, with the intention of having a single set of queues, but that work is incomplete. So the message needs to go into one of these queues and block_msgs is a fairly arbitrary choice.

Reply:
Heh.  That's ... both :+1: and :-1:.  Thanks.